contracts/services/osmosis-cl-lper/src/contract.rs

```diff
@@ -0,0 +1,232 @@
+        pool_id: cfg.lp_config.pool_id.u64(),
+        sender: cfg.input_addr.to_string(),
+        lower_tick: tick_range.lower_tick.i64(),
+        upper_tick: tick_range.upper_tick.i64(),
+        tokens_provided: cosmwasm_to_proto_coins(vec![bal_asset_1, bal_asset_2]),
+        token_min_amount0: token_min_amount_0.to_string(),
+        token_min_amount1: token_min_amount_1.to_string(),
+    }
+    .into();
+
+    // we delegate the create position msg as a submsg as we will need to transfer
+    // the position afterwards. reply_always so that the saved state is cleared on error.
+    let delegated_input_acc_msgs = execute_submsgs_on_behalf_of(
+        vec![SubMsg::reply_always(create_cl_position_msg, REPLY_ID)],
+        Some(to_json_string(&cfg)?),
+        &cfg.input_addr.clone(),
+    )?;
+
+    let service_submsg = SubMsg::reply_on_success(delegated_input_acc_msgs, REPLY_ID);
+
+    Ok(Response::default().add_submessage(service_submsg))
+}
+
+pub fn provide_liquidity_default(
+    deps: DepsMut,
+    cfg: Config,
+    bucket_count: Uint64,
+) -> Result<Response, ServiceError> {
+    // first we assert the input account balances
+    let bal_asset_1 = deps
+        .querier
+        .query_balance(&cfg.input_addr, cfg.lp_config.pool_asset_1.as_str())?;
+    let bal_asset_2 = deps
+        .querier
+        .query_balance(&cfg.input_addr, cfg.lp_config.pool_asset_2.as_str())?;
+
+    // query the pool config
+    let pool_cfg = query_cl_pool(&deps, cfg.lp_config.pool_id.u64())?;
+
+    // we derive the tick range from the bucket count
+    let active_bucket = TickRange::from(pool_cfg);
+
+    // we extend the active bucket range to both sides by the bucket count
+    let derived_tick_range = active_bucket.amplify_range_bidirectionally(bucket_count)?;
+
+    // the target range must be contained within the global tick range
+    cfg.lp_config
+        .global_tick_range
+        .ensure_contains(&derived_tick_range)?;
+
+    let create_cl_position_msg: CosmosMsg = MsgCreatePosition {
+        pool_id: cfg.lp_config.pool_id.u64(),
+        sender: cfg.input_addr.to_string(),
+        lower_tick: derived_tick_range.lower_tick.i64(),
+        upper_tick: derived_tick_range.upper_tick.i64(),
+        tokens_provided: cosmwasm_to_proto_coins(vec![bal_asset_1, bal_asset_2]),
+        token_min_amount0: "0".to_string(),
+        token_min_amount1: "0".to_string(),
+    }
+    .into();
+
+    // we delegate the position creation message to the input account
+    let delegated_input_acc_msgs = execute_submsgs_on_behalf_of(
+        // we expect a reply from this submsg so we pass it as a submessage
+        vec![SubMsg::reply_always(create_cl_position_msg, REPLY_ID)],
+        // associate this msg with a cfg payload which will be used in the reply
+        // to restore the state used during this function
+        Some(to_json_string(&cfg)?),
+        &cfg.input_addr.clone(),
+    )?;
```

Comment:
Why reply_always here?. Unless in the base account you want to save something when it errors (don't think it's the case as they don't store state afaik) this should be a reply_on_success. This way if the CL position fails we just stop

Reply:
good one; I had it there from before accounts had callbacks and I was clearing the temp state on both success & failure but this is indeed cleaner now